packages/openapi/src/helpers.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  getFriendlyName as getAssignedFriendlyName,
+  ModelType,
+  ModelTypeProperty,
+  Program,
+  Type,
+  TypeNameOptions,
+} from "@cadl-lang/compiler";
+import { reportDiagnostic } from "./lib.js";
+
+/**
+ * Determines whether a type will be inlined in OpenAPI rather than defined
+ * as a schema and referenced.
+ *
+ * All anonymous types (anonymous models, arrays, tuples, etc.) are inlined.
+ *
+ * Template instantiations are inlined unless they have a friendly name.
+ *
+ * A friendly name can be provided by the user using `@friendlyName`
+ * decorator, or chosen by default in simple cases.
+ */
+export function shouldInline(program: Program, type: Type): boolean {
+  if (hasFriendlyName(program, type)) {
+    return false;
+  }
+
+  switch (type.kind) {
+    case "Model":
+      return !type.name || hasTemplateArguments(type);
+    case "Enum":
+    case "Union":
+      return !type.name;
+    default:
+      return true;
+  }
+}
+
+/**
+ * Gets the name of a type to be used in OpenAPI.
+ *
+ * For inlined types: this is the Cadl-native name written to `x-cadl-name`.
+ *
+ * For non-inlined types: this is either the friendly name or the Cadl-native name.
+ *
+ * Cadl-native names are shortened to exclude root `Cadl` namespace and service
+ * namespace using the provided `TypeNameOptions`.
+ */
+export function getTypeName(
+  program: Program,
+  type: Type,
+  options: TypeNameOptions,
+  existing?: Record<string, any>
+): string {
+  const name =
+    getFriendlyName(program, type, options) ?? program.checker!.getTypeName(type, options);
+
+  if (existing && existing[name] !== undefined) {
+    reportDiagnostic(program, {
```

Comment:
Should we report both instances conflicting. That's what we have for duplicate-symbol or duplicate routes.